File: qm_cap.m
% Constants
e = 1.602e-19; % electron charge
k_B = 1.38064852e-23; % Boltzmann constant
T = 300; % temperature in Kelvin
mass_mos2 = 2.12488e-21;
mass_nc = 3.18161e-21;

% Density of states (DOS) calculated from DFT


% Define the thermal broadening function
thermal_broadening = @(E, Vex, T) sech((e*(E-Vex))/(2*k_B*T)).^2;

% Define the range of external potentials
Vext_range = -0.5:0.01:0.5;

% Initialize an array to store the capacitance values
Cq = zeros(size(Vext_range));

% Sweep the external potential
for i = 1:length(Vext_range)
    Vex = Vext_range(i);
    %Calculate the product of DOS and thermal broadening function
    dosF_T = nc_o_dos(:, 2) .* thermal_broadening(nc_o_dos(:, 1),Vex, T);
    % Define the integral function
    integrand = @(E) interp1(nc_o_dos(:, 1), dosF_T, E, 'spline');
    % Calculate the definite integral using integral
    Cq(i) = e^2*(4*k_B*T)^-1*integral(integrand, -inf, inf);
end

% Plot the results

plot(Vext_range, Cq/mass_nc);
xlabel('External Potential (V)');
ylabel('Quantum Capacitance (F/g)');

File: QM_Cap.py
''' In this code, we use NumPy for array manipulations, and SciPy's integrate.quad function to calculate the definite integral. 
    Make sure to replace 'dos_data.csv' with the correct file path to your CSV file containing the Density of States (DOS) data in two columns, 
    the first column representing energies and the second column representing DOS values. 
    The rest of the code should work as expected, plotting the Quantum Capacitance against the External Potential.
'''

import numpy as np
import pandas as pd
import scipy.constants as const
from scipy import integrate
import matplotlib.pyplot as plt

# Constants
e = const.elementary_charge
k_B = const.Boltzmann
T = 300  # temperature in Kelvin
mass_mos2 = 2.12488e-21
mass_nc = 3.18161e-21

# Read Density of States (DOS) from CSV file
dos_data = pd.read_csv('dos_data.csv')
nc_o_dos = dos_data.to_numpy()

# Define the thermal broadening function
def thermal_broadening(E, Vex, T):
    return np.power(1 / np.cosh((e * (E - Vex)) / (2 * k_B * T)), 2)

# Define the range of external potentials
Vext_range = np.arange(-0.5, 0.51, 0.01)

# Initialize an array to store the capacitance values
Cq = np.zeros(len(Vext_range))

# Sweep the external potential
for i, Vex in enumerate(Vext_range):
    # Calculate the product of DOS and thermal broadening function
    dosF_T = nc_o_dos[:, 1] * thermal_broadening(nc_o_dos[:, 0], Vex, T)
    # Define the integral function
    integrand = lambda E: np.interp(E, nc_o_dos[:, 0], dosF_T, left=0, right=0)
    # Calculate the definite integral using quad function from SciPy
    Cq[i], _ = integrate.quad(integrand, -np.inf, np.inf)

Cq *= e**2 * (4 * k_B * T)**-1

# Plot the results
plt.plot(Vext_range, Cq / mass_nc)
plt.xlabel('External Potential (V)')
plt.ylabel('Quantum Capacitance (F/g)')
plt.show()
